### PR TITLE
Fix unable to find root namespace

### PR DESCRIPTION
**Pull request by shyamd**
This fixes an issue finding the relative path when the root package is a namespace package. This is likely to break if the namespace package is nested in the module like `my_package.plugins` and you want to document something that lives in that plugins namespace.

### CODE REVIEW

File: src/pytkdocs/objects.py [hunk 1]
```diff
@@ -173,20 +173,26 @@ def relative_file_path(self) -> str:
...
+        for namespace in namespaces:
             try:
-                importlib.import_module(top_package_name)
-            except ImportError:
+                top_package = sys.modules[namespace]
+            except KeyError:
+                try:
+                    importlib.import_module(namespace)
+                except ImportError:
+                    return ""
```
Comment (pawamoy) [inline src/pytkdocs/objects.py:187]: We return early because if we couldn't import that namespace, we are sure that we could not import the subspace neither, right?

File: src/pytkdocs/objects.py [hunk 2]
```diff
@@ -173,20 +173,26 @@ def relative_file_path(self) -> str:
...
+                top_package = sys.modules[namespace]
+            try:
+                top_package_path = Path(inspect.getabsfile(top_package)).parent
+                return str(Path(self.file_path).relative_to(top_package_path.parent))
+            except ValueError:
                 return ""
```
Comment (pawamoy) [inline src/pytkdocs/objects.py:193]: Triggered by `Path.relative_to`, right? Same as above, we return early because if this path is not relative to the top package's parent path, then the subpath will not be relative to it either, right?

File: src/pytkdocs/objects.py [hunk 3]
```diff
@@ -173,20 +173,26 @@ def relative_file_path(self) -> str:
...
-            top_package = sys.modules[top_package_name]
-        top_package_path = Path(inspect.getabsfile(top_package)).parent
-        try:
-            return str(Path(self.file_path).relative_to(top_package_path.parent))
-        except ValueError:
-            return ""
+            except TypeError:
+                pass
```
Comment (pawamoy) [inline src/pytkdocs/objects.py:195]: Triggered by `inspect.getabsfile`? If this package didn't have an `__init__.py` file, we continue searching into subspaces?

File: src/pytkdocs/objects.py [hunk 4]
```diff
@@ -173,20 +173,26 @@ def relative_file_path(self) -> str:
...
```
Comment (pawamoy) [inline src/pytkdocs/objects.py:196]: Shouldn't we `return ""` once we're out of the loop?

File: src/pytkdocs/objects.py [hunk 5]
```diff
@@ -173,20 +173,26 @@ def relative_file_path(self) -> str:
 
         If the relative file path cannot be determined, the value returned is `""` (empty string).
         """
-        top_package_name = self.path.split(".", 1)[0]
-        try:
-            top_package = sys.modules[top_package_name]
-        except KeyError:
+
+        subspace_count = len(self.path.split("."))
+        namespaces = [".".join(self.path.split(".", maxsplit=maxsplit)[:-1]) for maxsplit in range(1, subspace_count)]
```
Comment (pawamoy) [inline src/pytkdocs/objects.py:178]: So with `self.path == "a.b.c"`, `namespaces == ["a", "a.b"]`, right?

This would be simpler and more efficient I think:

```python
parts = self.path.split(".")
namespaces = [".".join(parts[:l]) for l in range(1, len(parts))]
```

File: src/pytkdocs/objects.py (hunk 2 above)
Comment (shyamd) [reply src/pytkdocs/objects.py:193]: I believe that is correct. This was taken from the previous implementation. My understanding of `Path` is that it will throw a `ValueError` if `relative_to` can't resolve the paths.

File: src/pytkdocs/objects.py (hunk 3 above)
Comment (shyamd) [reply src/pytkdocs/objects.py:195]: Yup!

File: src/pytkdocs/objects.py (hunk 4 above)
Comment (shyamd) [reply src/pytkdocs/objects.py:196]: Good catch!

File: src/pytkdocs/objects.py (hunk 5 above)
Comment (shyamd) [reply src/pytkdocs/objects.py:178]: Nice!

**Comment by shyamd** [comment]
The nested `try`/`except` always bothers me so I thought something like this might be a bit more readable:
```
        for namespace in namespaces:
            try:
                importlib.import_module(namespace)
                top_package = sys.modules[namespace]
            except (KeyError, ImportError):
                """
                Triggered if we can't import the package via importlib
                Or if package isn't findable by it's canonical name even after importing
                """
                return ""

            try:
                top_package_path = Path(inspect.getabsfile(top_package)).parent
                return str(Path(self.file_path).relative_to(top_package_path.parent))
            except ValueError:
                return ""
            except TypeError:
                pass
```

**Comment by pawamoy** [comment]
This is definitely more readable! We always enforce the module is imported, and I guess it's not an expensive operation when it's already loaded, so it's fine.

File: src/pytkdocs/objects.py [hunk 6]
```diff
@@ -179,20 +179,24 @@ def relative_file_path(self) -> str:
...
-                try:
-                    importlib.import_module(namespace)
-                except ImportError:
-                    return ""
-                top_package = sys.modules[namespace]
+            except (KeyError, ImportError):
+                """
+                Triggered if we can't import the package via importlib
+                Or if package isn't findable by it's canonical name even after importing
+                """
```
Comment (pawamoy) [inline src/pytkdocs/objects.py:188]: Nitpicks: I'd prefer comments over a triple-quoted string :sweat_smile: : Also small typo: it's -> its

**Comment by pawamoy** [comment]
Alright, this looks great. Anything more you'd like to add? Also, do you mind if I squash the commits? I'm using the angular style, to ease the generation of the changelog.

**Comment by shyamd** [comment]
I'm trying to figure out how to build a test fixture since namespace packages have to be installed to get them to be located properly.

**Comment by pawamoy** [comment]
Nice! The `TypeError` is not covered but I think I can add a test for it myself to get used to namespace packages. Ready to merge?

**Comment by shyamd** [comment]
Hmmm. Actually no, that should have been covered since i'm attempting to load a namespace directory. Let me play with test a bit more.

**Comment by shyamd** [comment]
Ok. That was just a bad test. I've updated it to actually check the `relative_file_path` and ensure it points to the `__init__.py` file of the subspace module. This is both a good test case and an edge case since the previous code was failing when referencing the top subspace module.

**Comment by shyamd** [comment]
I can setup pushing to codecov if you setup the account and add the codecov token to github secrets for this repo. I just need to know the name of the secret to reference it in the workflow.

**Comment by pawamoy** [comment]
Not a fan of codecov :confused: (generally not a fan of third-party web services)

I don't really mind coverage going up and down as long as it stays above 85%. In itself coverage rarelly means "good test suite" :smile: 

If we can get a badge from GitHub workflow to show coverage percent in README, it's good enough for me! If not, well... it's fine. Not really important.

Anyway, thank you very much @shyamd! I really appreciate your work on this.

File: src/pytkdocs/objects.py [hunk 7]
```diff
@@ -175,17 +175,17 @@ def relative_file_path(self) -> str:
         """
 
         parts = self.path.split(".")
-        namespaces = [".".join(parts[:l]) for l in range(1, len(parts))]
-
+        namespaces = [".".join(parts[:l]) for l in range(1, len(parts) + 1)]
+        # Itterate through all sub namespaces including the last incase its a module
```
Comment (pawamoy) [inline src/pytkdocs/objects.py:179]: Ah, some typos: "Iterate" and "in case"

**Comment by pawamoy** [comment]
Thanks again! :tada:
